### PR TITLE
[IMP] hr_timesheet,**: improve generic ux for timesheet

### PR DESCRIPTION
**Pull request by kcv-odoo**
**  = hr_timesheet_attendance,project_timesheet_holiday,sale_timesheet

Purpose of this PR to improve generic usage of
timesheet app.

So, in this PR done following change:

- change no content helper for reporting menu actions.
- change no content helper for attendance report menu action.
- make timesheet editable in batch.
- add suspicious duration filter in timesheet search view.
- add analytic tags in timesheet views only visible when analytic
  tag feature is enable from accounting settings.
- remove group group_no_one from section_leaves div.
- remove group group_no_one from timesheet group in leave type
  form view.
- improve search view.
- remove form view reference from project_id of timesheet.
- add a condition for decoration danger in o2m tree view for
  timesheet_ids in project form view.

task-2725525

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I have made some comments. 🙂

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 1]
```diff
@@ -103,6 +105,12 @@ def _compute_user_id(self):
         for line in self:
             line.user_id = line.employee_id.user_id if line.employee_id else line._default_user()
 
+    @api.depends('employee_id')
+    def _compute_manager_id(self):
+        for line in self:
+            line.manager_id = line.employee_id.parent_id
```
[the remark this answers is not in the document]
Comment (kcv-odoo) [reply addons/hr_timesheet/models/hr_timesheet.py]: Done.

File: addons/hr_timesheet/static/src/js/task_with_hours.js [hunk 2]
```diff
@@ -38,7 +38,9 @@ const TaskWithHours = FieldMany2One.extend({
      * @private
      */
     _renderEdit: function (){
-        this.m2o_value = this._getDisplayNameWithoutHours(this.m2o_value);
+        if (this.m2o_value) {
+            this.m2o_value = this._getDisplayNameWithoutHours(this.m2o_value);
+        }
```
[the remark this answers is not in the document]
Comment (kcv-odoo) [reply addons/hr_timesheet/static/src/js/task_with_hours.js:43]: I add this because after enabling task creation from timer sometime it get undefined m2o_value [image](https://www.awesomescreenshot.com/image/21865994?key=26cf1d6db35d450d873b5c7da10bf776) that's I added this.

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 3]
```diff
@@ -43,12 +43,14 @@ def _domain_task_id(self):
         'project.task', 'Task', index='btree_not_null',
         compute='_compute_task_id', store=True, readonly=False,
         domain="[('company_id', '=', company_id), ('project_id.allow_timesheets', '=', True), ('project_id', '=?', project_id)]")
+    parent_task_id = fields.Many2one('project.task', related='task_id.parent_id', store=True)  # store for group by
```
Comment (xavierbol) [inline addons/hr_timesheet/models/hr_timesheet.py:46]: You can remove the comment. 🙂

File: addons/hr_timesheet/static/src/js/task_with_hours.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/hr_timesheet/static/src/js/task_with_hours.js:43]: I don't think it is that the problem because if `this.m2o_value = undefined` then `value && value.split('\u00A0')[0]` should return the value in `value` since we check `value` before using the `split` method. 🤔 
So, I don't think your changes fixed the issue. 🤔

**Review by xavierbol — COMMENTED** [review]
Thanks for the changes, I have made 2 comments. 🙂

File: addons/hr_timesheet/static/src/js/task_with_hours.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (kcv-odoo) [reply addons/hr_timesheet/static/src/js/task_with_hours.js]: Done. it's an issue from the `_getDisplayName` method so I overwrite that method and check `value` before using the `split` method.

File: addons/hr_timesheet/static/src/js/task_with_hours.js [hunk 4]
```diff
@@ -19,6 +19,14 @@ const TaskWithHours = FieldMany2One.extend({
...
         return value && value.split('\u00A0')[0];
     },
+    /**
+    * @private
+    * @param {string} value
+    * @returns {string} display_name
+    */
+    _getDisplayName: function (value) {
+        return value && value.split('\n')[0];
+    },
```
Comment (xavierbol) [inline addons/hr_timesheet/static/src/js/task_with_hours.js:29]: Why did you need to add this method?

Comment (kcv-odoo) [reply addons/hr_timesheet/static/src/js/task_with_hours.js:29]: I add this method because _onInputKeyup method of web calls _getDisplayName [line](https://github.com/odoo/odoo/blob/master/addons/web/static/src/legacy/js/fields/relational_fields.js#L870) After enabling task creation from timer sometimes it gets undefined m2o_value and [_getDisplayName](https://github.com/odoo/odoo/blob/master/addons/web/static/src/legacy/js/fields/relational_fields.js#L410) of web don't check the value before calling split function and throws an [error](https://www.awesomescreenshot.com/image/21865994?key=26cf1d6db35d450d873b5c7da10bf776) that's why I add this method and check value before calling split function. Should I check the value in the web _getDisplayName method instead of adding this method?

Comment (xavierbol) [reply addons/hr_timesheet/static/src/js/task_with_hours.js:29]: Sorry, but I don't understand why `this.m2o_value` is something equal to undefined. 🤔

Comment (kcv-odoo) [reply addons/hr_timesheet/static/src/js/task_with_hours.js]: Timer header task field create depending arguments of Timer when we start the timer we get undefined this.m2o_value and whenever the second argument of the timer is undefined to fix that I add an empty string for taskID field when there is no second argument in timer [commit](https://github.com/odoo/enterprise/pull/23284/commits/5cf0d47342e22304905c69c30064075c427381e1) and remove this function.

Comment (xavierbol) [reply addons/hr_timesheet/static/src/js/task_with_hours.js]: I cannot reproduce the bug. Can you share me a video?

Comment (kcv-odoo) [reply addons/hr_timesheet/static/src/js/task_with_hours.js]: I don't know the functional cause of this bug but sometimes it's reproducible [Video](https://rb.gy/bm6xfa) and most time it's not. :confused:

Comment (xavierbol) [reply addons/hr_timesheet/static/src/js/task_with_hours.js]: Thanks 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks for your work. 🙂

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 5]
```diff
@@ -43,12 +43,14 @@ def _domain_task_id(self):
         'project.task', 'Task', index='btree_not_null',
         compute='_compute_task_id', store=True, readonly=False,
         domain="[('company_id', '=', company_id), ('project_id.allow_timesheets', '=', True), ('project_id', '=?', project_id)]")
+    parent_task_id = fields.Many2one('project.task', related='task_id.parent_id', store=True)
```
Comment (IT-Ideas) [inline addons/hr_timesheet/models/hr_timesheet.py:46]: I am not sure that parent_task_id makes really sense as when having a situation with sub task of sub task ...
I am unsure have the level just above is really interesting. What would be best is to have the top level task. But this would imply storing the top level on the task (through a stored compute) and storing here the related which seems overkill to me.

I will pushback the PO and let you know 🙂

**Review by IT-Ideas — COMMENTED** [review]
Thanks for your work.
I will just push back the PO regarding the parent_task_id as I do not find this relevant as odoo is not limite to 2 level of tasks. We can have sub task of sub tasks.

**Comment by IT-Ideas** [comment]
Hello @kcv-odoo we finally agree that it would be best to store, search and group_by the `top level` task. We could maybe name it something like ancestor_task_id, maybe ask @aju-odoo if she has a better name.
I am still unsure what the best technical design would be best, just a recursive compute or trying to leverage the use of `_parent_store` 🤷🏼‍♂️. Maybe @xavierbol has an opinion on it.

**Comment by kcv-odoo** [comment]
> Hello @kcv-odoo we finally agree that it would be best to store, search and group_by the `top level` task. We could maybe name it something like ancestor_task_id, maybe ask @aju-odoo if she has a better name. I am still unsure what the best technical design would be best, just a recursive compute or trying to leverage the use of `_parent_store` 🤷🏼‍♂️. Maybe @xavierbol has an opinion on it.

Hello @IT-Ideas @xavierbol I added a commit to compute `top level` task. could you check? Thanks :slightly_smiling_face:

**Review by IT-Ideas — COMMENTED** [review]
I would have done it a bit differently.
What do you think of it?
Thanks for your work 🙂

File: addons/project/models/project.py [hunk 6]
```diff
@@ -1014,6 +1014,7 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
     legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing Explanation', readonly=True, related_sudo=False)
     is_closed = fields.Boolean(related="stage_id.fold", string="Closing Stage", store=True, index=True, related_sudo=False, help="Folded in Kanban stages are closing stages.")
     parent_id = fields.Many2one('project.task', string='Parent Task', index=True)
+    ancestor_task_id = fields.Many2one('project.task', compute='_compute_ancestor_task_id')
```
[the remark this answers is not in the document]
Comment (kcv-odoo) [reply addons/project/models/project.py]: Done!

File: addons/project/models/project.py [hunk 7]
```diff
@@ -1151,6 +1152,17 @@ def _compute_is_private(self):
...
+    @api.depends('parent_id', 'parent_id.parent_id')
+    def _compute_ancestor_task_id(self):
+        def _get_top_level_task(parent_task):
+            if parent_task.parent_id:
+                return _get_top_level_task(parent_task.parent_id)
+            return parent_task
+        no_parent_task = self.filtered(lambda t: not t.parent_id)
+        no_parent_task.ancestor_task_id = False
+        for task in (self - no_parent_task):
+            task.ancestor_task_id = _get_top_level_task(task.parent_id)
```
[the remark this answers is not in the document]
Comment (kcv-odoo) [reply addons/project/models/project.py]: No. :pensive:

**Review by IT-Ideas — APPROVED** [review]
Thanks for your work 🙂

File: addons/project/models/project.py [hunk 8]
```diff
@@ -1014,6 +1014,7 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
     legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing Explanation', readonly=True, related_sudo=False)
     is_closed = fields.Boolean(related="stage_id.fold", string="Closing Stage", store=True, index=True, related_sudo=False, help="Folded in Kanban stages are closing stages.")
     parent_id = fields.Many2one('project.task', string='Parent Task', index=True)
+    ancestor_task_id = fields.Many2one('project.task', compute='_compute_ancestor_task_id', index=True, recursive=True, store=True)
```
Comment (IT-Ideas) [inline addons/project/models/project.py:1017]: After having thought of it I would rather name it ancestor_id (as we are already in task model), what do you think?
Sorry for the change, I should have thought about it prior to give you the name 😕.

**Review by IT-Ideas — COMMENTED** [review]
One last thing.
You will also to make a migration script on order to pre-compute the ancestor_id.
Thanks!

File: addons/project/models/project.py (hunk 8 above)
Comment (kcv-odoo) [reply addons/project/models/project.py]: Done.

**Review by IT-Ideas — COMMENTED** [review]
Thanks 🙂 
An upgrade script to compute the ancestor_id and we will be good to go 🙂

**Comment by kcv-odoo** [comment]
> Thanks slightly_smiling_face An upgrade script to compute the ancestor_id and we will be good to go slightly_smiling_face

Hello @IT-Ideas done changes. :slightly_smiling_face:

File: addons/project/models/project.py [hunk 9]
```diff
@@ -1151,6 +1152,11 @@ def _compute_is_private(self):
         for task in self:
             task.is_private = not task.project_id and not task.parent_id
 
+    @api.depends('parent_id', 'parent_id.parent_id')
```
Comment (KangOl) [inline addons/project/models/project.py:1155]:
```suggestion
    @api.depends('parent_id', 'parent_id.ancestor_id')
```

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 10]
```diff
@@ -43,12 +43,14 @@ def _domain_task_id(self):
         'project.task', 'Task', index='btree_not_null',
         compute='_compute_task_id', store=True, readonly=False,
         domain="[('company_id', '=', company_id), ('project_id.allow_timesheets', '=', True), ('project_id', '=?', project_id)]")
+    ancestor_task_id = fields.Many2one('project.task', related='task_id.ancestor_id', store=True)
```
Comment (IT-Ideas) [inline addons/hr_timesheet/models/hr_timesheet.py:46]:
```suggestion
    ancestor_task_id = fields.Many2one('project.task', related='task_id.ancestor_id', store=True, index='btree_not_null')
```

File: addons/project/models/project.py [hunk 11]
```diff
@@ -1014,6 +1014,7 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
     legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing Explanation', readonly=True, related_sudo=False)
     is_closed = fields.Boolean(related="stage_id.fold", string="Closing Stage", store=True, index=True, related_sudo=False, help="Folded in Kanban stages are closing stages.")
     parent_id = fields.Many2one('project.task', string='Parent Task', index=True)
+    ancestor_id = fields.Many2one('project.task', compute='_compute_ancestor_id', index=True, recursive=True, store=True)
```
Comment (IT-Ideas) [inline addons/project/models/project.py:1017]:
```suggestion
    ancestor_id = fields.Many2one('project.task', compute='_compute_ancestor_id', index=`btree_not_null`,` recursive=True, store=True)
```

**Comment by IT-Ideas** [comment]
@kcv-odoo I wanted to rebase the branches on master. I realize that I fetched your dev branches but did forget I already had them locally so I lost your last changes. I think I found the missing piece but in case you have a doubt do not hesitate to go from your local branch, ignoring the remote one.

**Comment by KangOl** [comment]
❌ Some CI checks were not successful.

Can you please rebase your branches to ensure they are up to date?

File: addons/project/models/project.py [hunk 12]
```diff
@@ -1171,6 +1172,11 @@ def _compute_is_private(self):
         for task in self:
             task.is_private = not task.project_id and not task.parent_id
 
+    @api.depends('parent_id', 'parent_id.ancestor_id')
```
Comment (IT-Ideas) [inline addons/project/models/project.py:1175]: This is redundant, you can only specify `parent_id.ancestor_id` and it will do the same.
```suggestion
    @api.depends('parent_id.ancestor_id')
```

**Review by IT-Ideas — COMMENTED** [review]
Last small thing, then it looks good to me 🙂

File: addons/project/models/project.py (hunk 12 above)
Comment (kcv-odoo) [reply addons/project/models/project.py]: Done. :slightly_smiling_face:

**Review by IT-Ideas — APPROVED** [review]
Thanks for your work 🙂

**Comment by IT-Ideas** [comment]
@robodoo r+
